dataset/src/test/scala/frameless/functions/UdfTests.scala

```diff
@@ -4,27 +4,41 @@ package functions
 import org.scalacheck.Prop
 import org.scalacheck.Prop._
 
-class UdfTest extends TypedDatasetSuite {
+class UdfTests extends TypedDatasetSuite {
 
   test("one argument udf") {
-    def prop[A: TypedEncoder, B: TypedEncoder, C: TypedEncoder]
-    (data: Vector[X3[A, B, C]], f1: A => A): Prop = {
+    def prop[A: TypedEncoder, B: TypedEncoder](data: Vector[X1[A]], f1: A => B): Prop = {
       val dataset = TypedDataset.create(data)
-      val u1 = udf[X3[A, B, C], A, A](f1)
+      val u1 = udf[X1[A], A, B](f1)
       val u2 = dataset.makeUDF(f1)
       val A = dataset.col[A]('a)
 
-      val dataset21 = dataset.select(u1(A)).collect().run().toVector
-      val dataset22 = dataset.select(u2(A)).collect().run().toVector
+      // filter forces whole codegen
+      val codegen = dataset.filter(_ => true).select(u1(A)).collect().run().toVector
+
+      // otherwise it uses local relation
+      val local = dataset.select(u2(A)).collect().run().toVector
+
       val d = data.map(x => f1(x.a))
 
-      (dataset21 ?= d) && (dataset22 ?= d)
+      (codegen ?= d) && (local ?= d)
     }
 
-    check(forAll(prop[Int, Int, Int] _))
-    check(forAll(prop[String, Int, Int] _))
-    check(forAll(prop[Option[Int], X2[Double, Long], Int] _))
```

Comment:
According to implementation details, there shouldn't be a difference between `X1` and `X2` because they are both encoded using the same type of encoder. I wanted to simplify things a bit.